File: languages/typescript/extensions/content/handler.ts
/**
 * CONTENT §6 — the system content handler. Two operations, `get` (§6.2) and
 * `ingest` (§6.3), both hash-addressed at the params level and path-addressed at
 * the cap-scope level (§6.4).
 *
 * The handler is **optional** as an extension (§11.3) and type-agnostic by design
 * (§6): it serves any entity in the content store, not just blobs and chunks.
 */

import {
  Ecf,
  Entity,
  Envelope,
  HandlerResult,
  Status,
  errorResult,
  type ContentStore,
  type Handler,
  type HandlerContext,
  type HandlerOperations,
} from "entity-core-protocol-typescript";

import { ContentTypes, CONTENT_PATTERN } from "./types.js";

/**
 * Envelope + response overhead reserved out of the frame budget before entities are
 * packed. The response entity itself carries two hash arrays (33 B each plus CBOR
 * framing) and the envelope adds its own map framing; 4 KiB is comfortably above
 * both for any batch a caller can request within one frame.
 */
const FRAME_RESERVE_BYTES = 4096;

export interface ContentHandlerOptions {
  /**
   * The namespace prefix this handler instance serves (§6.4). `system/content` is
   * the default-namespace value §6.2 pins for the §6.1 manifest registration.
   *
   * **Topology.** Cycle 1 runs single-trust-domain (§6.4.1): `get` resolves any
   * hash in the content store and `ingest` writes the store without binding into
   * the tree. §6.4.1 makes that OPT-IN AND RESTRICTED — "MUST NOT be enabled as the
   * default configuration", and multi-party deployments running it are "out-of-spec
   * and security-defective". It is declared in `EXTENSION.toml [assumptions]
   * .topology` rather than assumed here, and the namespace-scoped topology (§6.4.2
   * Hash Tree Presence) is a declared cycle-1 gap, not a silent one.
   */
  readonly namespace?: string;
}

export class ContentHandler implements Handler {
  readonly pattern = CONTENT_PATTERN;
  readonly name = "content";

  /**
   * §6.1's manifest, declared here and published verbatim by `registerHandler`.
   *
   * This was `readonly string[]` for exactly one day. keystone's `Handler.operations`
   * carried operation NAMES only, so registration rendered `{get: {}, ingest: {}}` and
   * `installContent` had to re-write the interface entity afterwards to say what the
   * operations take and return. Routed as K-2; closed 2026-09-06 — and the sharper
   * version of the finding was theirs, not ours: the **wire** register op had always
   * forwarded a full §3.7 manifest, so the in-process surface was the narrower of the
   * two rather than both being narrow. The bare-name form still renders byte-identically,
   * so no bootstrap handler moved.
   */
  readonly operations: HandlerOperations = {
    get: { inputType: ContentTypes.GetRequest, outputType: ContentTypes.ContentResponse },
    ingest: { inputType: ContentTypes.IngestRequest, outputType: ContentTypes.IngestResult },
  };

  readonly #namespace: string;

  constructor(options: ContentHandlerOptions = {}) {
    this.#namespace = options.namespace ?? CONTENT_PATTERN;
  }

  async handle(ctx: HandlerContext): Promise<HandlerResult> {
    // §6.2 / §6.3 path-as-resource MUST. Checked FIRST, before the operation
    // switch, because it applies to both ops identically and because an unknown
    // operation with no resource should still report the resource failure a caller
    // can act on. `400 path_required` — the status is pinned by
    // GUIDE-EXTENSION-DEVELOPMENT §171, and the oracle checks only the CODE, so a
    // peer answering 404 passes the gate and fails the spec. We answer 400.
    if (ctx.resource === null) {
      return errorResult(
        Status.BadRequest,
        "path_required",
        `${CONTENT_PATTERN}:${ctx.operation} requires a resource naming the namespace path (§6.2/§6.3)`,
      );
    }

    // §6.4 step 2 — path-scope check, handler-level. Dispatch already verified the
    // resource against the grant's `resources` scope (§6.4 step 1), so this is
    // defence-in-depth: it refuses a target outside the namespace this instance
    // serves even if a grant somehow covered it.
    //
    // `capability_denied`, NOT `forbidden`. v3.7 corrected §6.4's pseudocode: `forbidden`
    // was defined in no code set — it is ENTITY-CORE-PROTOCOL §3.3's *fallback* for the
    // 403 status, not a code — and §3.3's 403 row names `capability_denied` as the default.
    const outside = ctx.resource.targets.find((t) => !withinNamespace(t, this.#namespace));
    if (outside !== undefined) {
      return errorResult(
        Status.Forbidden,
        "capability_denied",
        `resource target '${outside}' is outside namespace '${this.#namespace}' (§6.4)`,
      );
    }

    switch (ctx.operation) {
      case "get":
        return this.#get(ctx);
      case "ingest":
        return this.#ingest(ctx);
      default:
        // §6.6(2): "This spec does not define a `system/content:delete`." Removal is
        // local GC, never a protocol op — so an unknown verb is 501, not 404.
        return errorResult(Status.NotSupported, "unsupported_operation", ctx.operation);
    }
  }

  // ── §6.2 get ────────────────────────────────────────────────────────────────

  #get(ctx: HandlerContext): HandlerResult {
    let hashes: readonly Uint8Array[];
    try {
      hashes = Ecf.asArray(Ecf.require(ctx.params.data, "hashes")).map((h) => Ecf.asBytes(h));
    } catch {
      return errorResult(Status.BadRequest, "unexpected_params", "get expects {hashes: array_of system/hash} (§6.2)");
    }

    const store: ContentStore = ctx.peer.contentStore;

    // Amendment 1 §6.2 MUST: consult THE CONNECTION'S configured budget at
    // response-construction time, not a hardcoded 16 MiB literal. `frameBudget()`
    // prefers `connection.maxFrameBytes` and falls back to the peer default only
    // for an in-process dispatch, which has no connection and still needs a number.
    let remaining = ctx.frameBudget() - FRAME_RESERVE_BYTES;

    const found: Uint8Array[] = [];
    const missing: Uint8Array[] = [];
    const included: Entity[] = [];
    // Once the budget is exhausted every REMAINING hash goes to `missing` in
    // request order — §6.2 says "as many as fit (in request order)", so a small
    // entity after a large one does NOT get packed. Order is the contract; the
    // requester retries with `missing` and makes progress deterministically.
    let budgetExhausted = false;

    for (const hash of hashes) {
      const entity = store.get(hash);
      if (entity === undefined) {
        missing.push(hash);
        continue;
      }
      const cost = entity.wireBytes.length + hash.length;
      if (budgetExhausted || cost > remaining) {
        budgetExhausted = true;
        missing.push(hash);
        continue;
      }
      remaining -= cost;
      included.push(entity);
      found.push(hash);
    }

    // §6.2 Amendment 2: `pending` is OPTIONAL and SHOULD be populated only by an
    // implementation with sync-state visibility — an active subscription on the
    // namespace plus an inbox feeding the content store. This composition has
    // neither, so the field is OMITTED. Emitting an empty array would advertise a
    // capability we do not have; §6.2 says a receiver that omits it is telling the
    // caller to treat all `missing` as terminal, which is the truth here.
    const response = Entity.create(
      ContentTypes.ContentResponse,
      Ecf.map(
        ["found", Ecf.array(found.map((h) => Ecf.bytes(h)))],
        ["missing", Ecf.array(missing.map((h) => Ecf.bytes(h)))],
      ),
    );
    return HandlerResult.ok(response, included);
  }

  // ── §6.3 ingest ─────────────────────────────────────────────────────────────

  #ingest(ctx: HandlerContext): HandlerResult {
    const envelopeValue = Ecf.field(ctx.params.data, "envelope");
    const entityValue = Ecf.field(ctx.params.data, "entity");

    if (envelopeValue !== null && entityValue !== null) {
      return errorResult(Status.BadRequest, "ambiguous_input", "Specify envelope or entity, not both");
    }
    if (envelopeValue === null && entityValue === null) {
      return errorResult(Status.BadRequest, "missing_input", "Specify envelope or entity");
    }

    const store: ContentStore = ctx.peer.contentStore;

    // ── Entity mode: store a single entity. `root` is ABSENT from the result —
    // there is no envelope wrapper to pass through (§6.3), and the §11.1 MUST is
    // scoped to envelope mode with a non-null root.
    if (entityValue !== null) {
      let entity: Entity;
      try {
        entity = Entity.fromDecoded(entityValue);
      } catch (err) {
        return errorResult(Status.BadRequest, "unexpected_params", `entity: ${message(err)}`);
      }
      store.put(entity);
      return HandlerResult.ok(
        Entity.create(
          ContentTypes.IngestResult,
          Ecf.map(["root_hash", Ecf.bytes(entity.contentHash)], ["ingested_count", Ecf.uint(1n)]),
        ),
      );
    }

    // ── Envelope mode: store root + all included.
    //
    // `system/envelope` extends `core/envelope`, whose `root` is a REQUIRED
    // `core/entity`, so §6.3's `if envelope.root is not null` branch is unreachable
    // for a well-formed envelope and a missing root is a malformed request rather
    // than an empty-count success. Recorded in the authoring notes; not routed,
    // because the type declaration already answers it.
    let envelope: Envelope;
    try {
      envelope = Envelope.decode(Ecf.encodeEcf(envelopeValue as NonNullable<typeof envelopeValue>));
    } catch (err) {
      // `Envelope.decode` is where the §6.3 hash-validation MUST is enforced: it
      // rejects an included entry whose content hash does not match its map key
      // (core §3.1). Reusing it rather than re-implementing the check is the point
      // — a second implementation of "does this hash match" is a second thing that
      // can disagree with the peer about entity fidelity.
      return errorResult(Status.BadRequest, "hash_mismatch", `envelope: ${message(err)}`);
    }

    store.put(envelope.root);
    let count = 1;
    for (const entity of envelope.included.values()) {
      store.put(entity);
      count++;
    }

    // §11.1 MUST: `root` is included, inlined, in envelope mode. It lets a
    // continuation navigate `data.root.data.<field>` without dereferencing the
    // content store (§6.3.1).
    return HandlerResult.ok(
      Entity.create(
        ContentTypes.IngestResult,
        Ecf.map(
          ["root", Ecf.decodeEcf(envelope.root.wireBytes)],
          ["root_hash", Ecf.bytes(envelope.root.contentHash)],
          ["ingested_count", Ecf.uint(BigInt(count))],
        ),
      ),
    );
  }
}

/** A resource target is in-namespace if it IS the prefix or sits under it. */
function withinNamespace(target: string, namespace: string): boolean {
  const t = target.replace(/^\/+/, "");
  return t === namespace || t.startsWith(namespace + "/");
}

function message(err: unknown): string {
  return err instanceof Error ? err.message : String(err);
}

File: languages/rust/extensions/content/tests/handler.rs
//! §6.2 / §6.3 handler behaviour, driven against a real peer `Store`.
//!
//! **These tests are ours.** They drive the body directly through [`HandlerRequest`], because a
//! third party cannot construct the peer's `HandlerContext`. The oracle's `content` category
//! reaches the same body over the wire now that `install_content` puts it behind
//! `Peer::register_handler`; what these add is the branches no vector reaches.

use entity_content::{
    create_blob_fixed, store_blob, ContentHandler, FrameBudget, HandlerRequest, CONTENT_PATTERN,
    CONTENT_RESPONSE, INGEST_RESULT,
};
use entity_core_protocol::peer::model::{self, Entity};
use entity_core_protocol::peer::store::Store;
use entity_core_protocol::value::{Key, Value};

fn targets(paths: &[&str]) -> Value {
    Value::Map(vec![(
        Key::Text("targets".into()),
        Value::Array(paths.iter().map(|p| Value::Text(p.to_string())).collect()),
    )])
}

fn exec(operation: &str, params: Entity, resource: Option<Value>) -> Entity {
    let mut pairs: Vec<(Key, Value)> = vec![
        (Key::Text("request_id".into()), model::text("t1")),
        (Key::Text("uri".into()), model::text(CONTENT_PATTERN)),
        (Key::Text("operation".into()), model::text(operation)),
        (Key::Text("params".into()), params.to_cbor()),
    ];
    if let Some(r) = resource {
        pairs.push((Key::Text("resource".into()), r));
    }
    Entity::make("system/protocol/execute", Value::Map(pairs))
}

fn get_params(hashes: &[Vec<u8>]) -> Entity {
    Entity::make(
        "system/content/get-request",
        Value::Map(vec![(
            Key::Text("hashes".into()),
            Value::Array(hashes.iter().map(|h| Value::Bytes(h.clone())).collect()),
        )]),
    )
}

fn req<'a>(e: &'a Entity, store: &'a Store) -> HandlerRequest<'a> {
    HandlerRequest {
        exec: e,
        store,
        frame_budget: FrameBudget::Enforced(entity_core_protocol::peer::wire::MAX_FRAME),
        caller_capability: None,
        local_peer: "PEER",
    }
}

fn code_of(result: &Entity) -> String {
    result.text_field("code").unwrap_or("").to_string()
}

// ── §6.2 / §6.3 the path-as-resource MUST ────────────────────────────────────

#[test]
fn a_request_with_no_resource_is_400_path_required() {
    // The status is pinned at 400 by GUIDE-EXTENSION-DEVELOPMENT §171. The oracle
    // checks only the CODE, so a peer answering 404 passes the gate and fails the
    // spec -- which is why the status is asserted here and not only the code.
    let store = Store::new();
    let h = ContentHandler::new();
    let e = exec("get", get_params(&[]), None);
    let out = h.handle_op("get", &req(&e, &store));
    assert_eq!(out.status, 400);
    assert_eq!(code_of(&out.result), "path_required");
}

#[test]
fn an_empty_targets_list_is_treated_as_absent() {
    // Core §3.2 makes `targets` MUST-contain-at-least-one, so `{targets: []}` is
    // malformed. See the note on `resource_targets`: over the wire, `check_permission`
    // refuses this first with 403 (measured on `python`). This test is an in-process call, where 400 is the
    // answer -- the same branch `python` keeps for its own in-process case.
    let store = Store::new();
    let h = ContentHandler::new();
    let e = exec("get", get_params(&[]), Some(targets(&[])));
    let out = h.handle_op("get", &req(&e, &store));
    assert_eq!(out.status, 400);
    assert_eq!(code_of(&out.result), "path_required");
}

#[test]
fn a_target_outside_the_namespace_is_403() {
    let store = Store::new();
    let h = ContentHandler::new();
    let e = exec("get", get_params(&[]), Some(targets(&["system/tree/x"])));
    let out = h.handle_op("get", &req(&e, &store));
    assert_eq!(out.status, 403);
    // v3.7 §6.4: `capability_denied`, not `forbidden`. Asserted on the CODE and not on the
    // status alone, because the status was already right under v3.6 and the code was not —
    // a test that checked only `403` would have survived the re-pin without noticing.
    assert_eq!(code_of(&out.result), "capability_denied");
}

/// §6.4 step 2 — the PATH-SCOPE check against the caller's capability, through the peer's H9
/// `check_path_permission` with handler pattern `system/content`. A target INSIDE the namespace (so
/// the prefix check above passes) that the grant does not cover is `403 capability_denied`. CONTROL:
/// the same request under a grant that does cover it proceeds — without it, a handler denying every
/// capability-bearing request would pass the first assertion.
#[test]
fn a_target_the_capability_does_not_cover_is_403_even_inside_the_namespace() {
    let store = Store::new();
    let h = ContentHandler::new();
    let grant = |resources: &str| {
        let scope = |s: &str| Value::Map(vec![(Key::Text("include".into()), Value::Array(vec![Value::Text(s.into())]))]);
        Entity::make(
            "system/capability/token",
            Value::Map(vec![(
                Key::Text("grants".into()),
                Value::Array(vec![Value::Map(vec![
                    (Key::Text("handlers".into()), scope("system/content")),
                    (Key::Text("resources".into()), scope(resources)),
                    (Key::Text("operations".into()), scope("get")),
                ])]),
            )]),
        )
    };
    let e = exec("get", get_params(&[]), Some(targets(&["system/content/private/x"])));
    let narrow = grant("system/content/public/*");
    let mut r = req(&e, &store);
    r.caller_capability = Some(&narrow);
    let out = h.handle_op("get", &r);
    assert_eq!((out.status, code_of(&out.result).as_str()), (403, "capability_denied"));

    let wide = grant("system/content/*");
    r.caller_capability = Some(&wide);
    assert_eq!(h.handle_op("get", &r).status, 200, "control: a covering grant proceeds");
}

/// §6.2 — the budget pays for the `found`/`missing` entries too, one per requested hash. A budget
/// that EXACTLY fits one entity plus one entry packs it (the CONTROL: the arithmetic is right, not
/// merely stingy); the same budget with one extra, absent hash in the request does not, because that
/// hash's `missing` entry is part of the response. The pre-2026-09-12 accounting packed it anyway.
#[test]
fn the_budget_is_charged_for_every_hash_list_entry_not_only_packed_entities() {
    let store = Store::new();
    let h = ContentHandler::new();
    let e = Entity::make("test/blob", Value::Map(vec![(Key::Text("p".into()), Value::Bytes(vec![7; 200]))]));
    store.put_entity(&e);
    let wire = entity_core_protocol::cbor::encode(&e.to_cbor()).len();
    let budget = entity_content::FRAME_RESERVE_BYTES + 35 + wire + e.hash.len();
    let run = |hashes: &[Vec<u8>]| {
        let x = exec("get", get_params(hashes), Some(targets(&[CONTENT_PATTERN])));
        let mut r = req(&x, &store);
        r.frame_budget = FrameBudget::Enforced(budget);
        let out = h.handle_op("get", &r);
        assert_eq!(out.status, 200);
        out.included.len()
    };
    assert_eq!(run(&[e.hash.clone()]), 1, "control: a budget that exactly fits packs the entity");
    let absent = vec![0u8; 33];
    assert_eq!(run(&[e.hash.clone(), absent]), 0, "the absent hash's `missing` entry is part of the response");
}

#[test]
fn an_unknown_operation_is_501_not_404() {
    // §6.6(2): "This spec does not define a system/content:delete." Removal is local
    // GC, never a protocol op.
    let store = Store::new();
    let h = ContentHandler::new();
    let e = exec("delete", get_params(&[]), Some(targets(&[CONTENT_PATTERN])));
    let out = h.handle_op("delete", &req(&e, &store));
    assert_eq!(out.status, 501);
    assert_eq!(code_of(&out.result), "unsupported_operation");
}

// ── §6.2 get ─────────────────────────────────────────────────────────────────

#[test]
fn get_returns_found_and_missing_as_arrays_and_includes_the_entities() {
    // `found` / `missing` are ARRAYS, not counters -- the F4 cross-impl audit landing,
    // and the one wire-shape mistake this response has already made once.
    let store = Store::new();
    let blob = create_blob_fixed(b"the quick brown fox", 8);
    let blob_hash = store_blob(&store, &blob);
    let absent = vec![0u8; 33];

    let h = ContentHandler::new();
    let e = exec(
        "get",
        get_params(&[blob_hash.clone(), absent.clone()]),
        Some(targets(&[CONTENT_PATTERN])),
    );
    let out = h.handle_op("get", &req(&e, &store));

    assert_eq!(out.status, 200);
    assert_eq!(out.result.typ, CONTENT_RESPONSE);
    match out.result.field("found") {
        Some(Value::Array(items)) => {
            assert_eq!(items.as_slice(), [Value::Bytes(blob_hash.clone())])
        }
        other => panic!("found is not an array: {other:?}"),
    }
    match out.result.field("missing") {
        Some(Value::Array(items)) => assert_eq!(items.as_slice(), [Value::Bytes(absent)]),
        other => panic!("missing is not an array: {other:?}"),
    }
    // §6.2's wire contract: the fetched entities are delivered via the response
    // envelope's `included` map, not inline in the result.
    assert_eq!(out.included.len(), 1);
    assert_eq!(out.included[0].hash, blob_hash);
}

#[test]
fn pending_is_omitted_and_not_emitted_empty() {
    // §6.2 Amendment 2. `absent` and `empty` mean the same thing to the mapping table,
    // but a reader diffing two peers would read an empty array as "supports sync-state,
    // nothing pending" -- which would advertise a capability this composition does not
    // have. Asserting ABSENCE is the only way this distinction is defended.
    let store = Store::new();
    let h = ContentHandler::new();
    let e = exec("get", get_params(&[]), Some(targets(&[CONTENT_PATTERN])));
    let out = h.handle_op("get", &req(&e, &store));
    assert!(
        out.result.field("pending").is_none(),
        "pending must be OMITTED, not emitted empty"
    );
}

#[test]
fn the_frame_budget_bounds_the_batch_and_the_rest_go_to_missing_in_order() {
    // §6.2: "as many as fit (in request order)". Once exhausted, every REMAINING hash
    // goes to `missing` -- a small entity after a large one does NOT get packed,
    // because order is the contract and the requester retries with `missing`.
    let store = Store::new();
    let big = create_blob_fixed(&vec![7u8; 4000], 4000);
    let big_hash = store_blob(&store, &big);
    let small = create_blob_fixed(b"x", 8);
    let small_hash = store_blob(&store, &small);

    let h = ContentHandler::new();
    let e = exec(
        "get",
        get_params(&[big_hash.clone(), small_hash.clone()]),
        Some(targets(&[CONTENT_PATTERN])),
    );
    // A budget deliberately too small for the first entity. `Enforced` is constructed
    // directly here rather than via `from_peer`, which is the whole reason the variant
    // carries its number: a peer that DID configure a budget would pass it the same way.
    let tight = HandlerRequest {
        exec: &e,
        store: &store,
        frame_budget: FrameBudget::Enforced(entity_content::FRAME_RESERVE_BYTES + 64),
        caller_capability: None,
        local_peer: "PEER",
    };
    let out = h.handle_op("get", &tight);
    assert_eq!(out.status, 200);
    match out.result.field("found") {
        Some(Value::Array(items)) => assert!(items.is_empty(), "nothing should have fit"),
        other => panic!("found is not an array: {other:?}"),
    }
    match out.result.field("missing") {
        Some(Value::Array(items)) => assert_eq!(
            items.as_slice(),
            [Value::Bytes(big_hash), Value::Bytes(small_hash)],
            "the small entity must NOT be packed after the large one was skipped"
        ),
        other => panic!("missing is not an array: {other:?}"),
    }
    assert!(out.included.is_empty());
}

/// Amendment 1 BY VALUE, which this peer could not support before keystone's H6: configure a
/// non-default budget and require the body's read to return it. The NEGATIVE arm is a peer left at
/// the default, whose read must be the 16 MiB literal — so a `from_peer` that returned the constant
/// would pass the second assertion and fail the first.
#[test]
fn from_peer_reads_the_configured_budget_by_value() {
    use entity_core_protocol::peer::{CreateOptions, Peer, PeerConfig};
    let opts = || CreateOptions { seed: [0x11; 32], open_grants: true, conformance: false };
    let configured = Peer::create_with(opts(), PeerConfig::default().max_frame_bytes(3_145_749));
    assert_eq!(FrameBudget::from_peer(&configured), FrameBudget::Enforced(3_145_749));
    let default = Peer::create(opts());
    assert_eq!(
        FrameBudget::from_peer(&default),
        FrameBudget::Enforced(entity_core_protocol::peer::wire::MAX_FRAME)
    );
}

// ── §6.3 ingest ──────────────────────────────────────────────────────────────

#[test]
fn ingest_entity_mode_stores_one_and_omits_root() {
    // §11.1's `root` MUST is scoped to ENVELOPE mode. In entity mode there is no
    // envelope wrapper to pass through, so `root` is absent -- which is exactly what
    // `optional` encodes under the §1.3 absent-key rule.
    let store = Store::new();
    let h = ContentHandler::new();
    let payload = Entity::make(
        "system/content/chunk",
        Value::Map(vec![(Key::Text("payload".into()), Value::Bytes(b"hi".to_vec()))]),
    );
    let params = Entity::make(
        "system/content/ingest-request",
        Value::Map(vec![(Key::Text("entity".into()), payload.to_cbor())]),
    );
    let e = exec("ingest", params, Some(targets(&[CONTENT_PATTERN])));
    let out = h.handle_op("ingest", &req(&e, &store));

    assert_eq!(out.status, 200);
    assert_eq!(out.result.typ, INGEST_RESULT);
    assert_eq!(out.result.uint_field("ingested_count"), Some(1));
    assert_eq!(out.result.bytes_field("root_hash"), Some(payload.hash.as_slice()));
    assert!(
        out.result.field("root").is_none(),
        "root MUST be absent in entity mode"
    );
    assert!(store.get_by_hash(&payload.hash).is_some());
}

#[test]
fn ingest_envelope_mode_inlines_root_and_counts_included() {
    let store = Store::new();
    let h = ContentHandler::new();
    let inner = Entity::make(
        "system/content/chunk",
        Value::Map(vec![(Key::Text("payload".into()), Value::Bytes(b"a".to_vec()))]),
    );
    let root = Entity::make(
        "system/content/blob",
        Value::Map(vec![
            (Key::Text("total_size".into()), Value::UInt(1)),
            (Key::Text("chunk_size".into()), Value::UInt(8)),
            (Key::Text("chunking".into()), Value::UInt(0)),
            (
                Key::Text("chunks".into()),
                Value::Array(vec![Value::Bytes(inner.hash.clone())]),
            ),
        ]),
    );
    let envelope = Value::Map(vec![
        (Key::Text("root".into()), root.to_cbor()),
        (
            Key::Text("included".into()),
            Value::Map(vec![(Key::Bytes(inner.hash.clone()), inner.to_cbor())]),
        ),
    ]);
    let params = Entity::make(
        "system/content/ingest-request",
        Value::Map(vec![(Key::Text("envelope".into()), envelope)]),
    );
    let e = exec("ingest", params, Some(targets(&[CONTENT_PATTERN])));
    let out = h.handle_op("ingest", &req(&e, &store));

    assert_eq!(out.status, 200);
    assert_eq!(out.result.uint_field("ingested_count"), Some(2));
    // §11.1 MUST: `root` is inlined so a continuation can navigate
    // `data.root.data.<field>` without dereferencing the content store (§6.3.1).
    assert!(out.result.field("root").is_some(), "root MUST be inlined");
    assert!(store.get_by_hash(&root.hash).is_some());
    assert!(store.get_by_hash(&inner.hash).is_some());
}

#[test]
fn an_included_key_that_does_not_match_its_entity_is_400_hash_mismatch() {
    // The §3.1 KEY check, which is a different assertion from `entity_of_cbor`'s
    // re-derivation of the hash from {type, data}, and it is the one §6.3 names.
    let store = Store::new();
    let h = ContentHandler::new();
    let inner = Entity::make(
        "system/content/chunk",
        Value::Map(vec![(Key::Text("payload".into()), Value::Bytes(b"a".to_vec()))]),
    );
    let root = Entity::make("system/content/blob", Value::Map(vec![]));
    let envelope = Value::Map(vec![
        (Key::Text("root".into()), root.to_cbor()),
        (
            Key::Text("included".into()),
            Value::Map(vec![(Key::Bytes(vec![0u8; 33]), inner.to_cbor())]),
        ),
    ]);
    let params = Entity::make(
        "system/content/ingest-request",
        Value::Map(vec![(Key::Text("envelope".into()), envelope)]),
    );
    let e = exec("ingest", params, Some(targets(&[CONTENT_PATTERN])));
    let out = h.handle_op("ingest", &req(&e, &store));
    assert_eq!(out.status, 400);
    assert_eq!(code_of(&out.result), "hash_mismatch");
}

#[test]
fn envelope_and_entity_together_is_400_ambiguous_and_neither_is_400_missing() {
    let store = Store::new();
    let h = ContentHandler::new();

    let both = Entity::make(
        "system/content/ingest-request",
        Value::Map(vec![
            (Key::Text("envelope".into()), Value::Map(vec![])),
            (Key::Text("entity".into()), Value::Map(vec![])),
        ]),
    );
    let e = exec("ingest", both, Some(targets(&[CONTENT_PATTERN])));
    let out = h.handle_op("ingest", &req(&e, &store));
    assert_eq!(out.status, 400);
    assert_eq!(code_of(&out.result), "ambiguous_input");

    let neither = Entity::make("system/content/ingest-request", Value::Map(vec![]));
    let e = exec("ingest", neither, Some(targets(&[CONTENT_PATTERN])));
    let out = h.handle_op("ingest", &req(&e, &store));
    assert_eq!(out.status, 400);
    assert_eq!(code_of(&out.result), "missing_input");
}

#[test]
fn an_envelope_with_no_root_is_400_not_an_empty_success() {
    // §6.3's `if envelope.root is not null` branch is unreachable for a well-formed
    // envelope: `system/envelope` extends `core/envelope`, whose `root` is a REQUIRED
    // `core/entity`. So a missing root is a malformed request, not an ingested_count:0.
    let store = Store::new();
    let h = ContentHandler::new();
    let params = Entity::make(
        "system/content/ingest-request",
        Value::Map(vec![(Key::Text("envelope".into()), Value::Map(vec![]))]),
    );
    let e = exec("ingest", params, Some(targets(&[CONTENT_PATTERN])));
    let out = h.handle_op("ingest", &req(&e, &store));
    assert_eq!(out.status, 400);
    assert_eq!(code_of(&out.result), "unexpected_params");
}
